### PR TITLE
Add term generator with tests

### PR DESCRIPTION
**Pull request by sebb7**
Module requires all previously implemented generators which can be found in last PRs.

### CODE REVIEW

File: lib/orangeade/generator/boolean_and_nil.ex [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+  """
+  @spec stream() :: Caffeine.Stream.t()
+  def stream do
+    Caffeine.Stream.map(BoundNatural.stream(limit: 3), &to_boolean_or_nil/1)
+  end
+
+  defp to_boolean_or_nil(n) do
+    case n do
+      0 -> nil
```
Comment (Dzol) [inline lib/orangeade/generator/boolean_and_nil.ex:27]: I hadn't thought of including `nil` together with a boolean generator (like I wrote here: https://github.com/lambdaacademy/orangeade/pull/8#discussion_r183048447). Your name for the generator is good.

File: lib/orangeade/generator/binary.ex [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  defp reduce_stream(0, reduced_stream) do
+    reduced_stream
+  end
+  defp reduce_stream(i, stream) do
+    reduce_stream(i - 1, Caffeine.Stream.tail(stream))
+  end
+
+  defp get_binary(list) do
+    :erlang.list_to_binary(list)
```
Comment (Dzol) [inline lib/orangeade/generator/binary.ex]: If you want to stick to Elixir as much as possible (w/o reaching into Erlang libraries) then you can go for [IO.iodata_to_binary/1](https://hexdocs.pm/elixir/IO.html#iodata_to_binary/1).

File: test/orangeade/generator/ascii_string_test.exs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+defmodule ASCIIStringTest do
+  use ExUnit.Case
+  alias Orangeade.Generator.ASCIIString
+
+  test "Check if first 10 generated elements are bitstrings" do
+    list_of_elements = ASCIIString.stream(max_word_length: 10)
+    |> Caffeine.Stream.take(10)
+    assert Enum.all?(list_of_elements, &is_bitstring/1)
```
Comment (Dzol) [inline test/orangeade/generator/ascii_string_test.exs]: There's `List.ascii_printable` (after conversion). `String.valid?` seems like the best choice for this assertion though.

File: test/orangeade/generator/term_test.exs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+defmodule Orangeade.Generator.TermTest do
+  use ExUnit.Case
+  alias Orangeade.Generator.Term
+
+  test "Check if there are different data types in first 50 generated elements" do
+    list_of_elements = Term.stream()
+    |> Caffeine.Stream.take(100)
+    list_of_conditions = [
+      Enum.any?(list_of_elements, &is_negative/1),
+      Enum.any?(list_of_elements, &is_in_range_32_126/1),
```
Comment (Dzol) [inline test/orangeade/generator/term_test.exs]: Why are we testing in the range between 32 and 126.

File: test/orangeade/generator/term_test.exs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+defmodule Orangeade.Generator.TermTest do
+  use ExUnit.Case
+  alias Orangeade.Generator.Term
+
+  test "Check if there are different data types in first 50 generated elements" do
```
Comment (Dzol) [inline test/orangeade/generator/term_test.exs]: This is meant to be 100 elements... isn't it?

File: test/orangeade/generator/term_test.exs (hunk 4 above)
Comment (Dzol) [reply test/orangeade/generator/term_test.exs]: Oh, okay, they're from the character stream. If you call is `ascii_printable?` or `is_ascii_printable?` then the name will convey what it means to be in the range from 32 to 126.

**Review by Dzol — COMMENTED** [review]
I like the Term generator a lot. I had thought of using `Enum.random/1` for selecting a generator out of the list of all generators but I like you solution more given the context.

I know a `get_` prefix is common in some languages + I see it in Elixir too but I always question why we name functions with a `get_*`. They don't get from anywhere... e.g. `get_binary/` just does a transformation. It doesn't get anything from anywhere.

There's one more change... writing `[_|_]` to extract pieces of the stream is convenient but their implementation may well change. The API could/should stay the same so using `Stream.{head,tail}/1` is important. The Caffeine stream "type" is opaque: https://hexdocs.pm/caffeine/Caffeine.Stream.html#t:t/0

**Comment by sebb7** [comment]
Changes added as required in the review. @Dzol, rebase was done so everything is ready to merge.

**Pull request merged**